pkg/log/context.go

```diff
@@ -74,40 +94,70 @@ func (s *Settings) Validate() error {
 	if len(s.Format) == 0 {
 		return fmt.Errorf("validate Settings: LogFormat missing")
 	}
-	if s.Output == nil {
+	if len(s.Output) == 0 {
 		return fmt.Errorf("validate Settings: LogOutput missing")
 	}
+
 	return nil
 }
 
 // Configure creates a new context with a logger using the provided settings.
 func Configure(ctx context.Context, settings *Settings) context.Context {
-	once.Do(func() {
-		level, err := logrus.ParseLevel(settings.Level)
-		if err != nil {
-			panic(fmt.Sprintf("Could not parse log level configuration: %s", err))
-		}
-		formatter, ok := supportedFormatters[settings.Format]
-		if !ok {
-			panic(fmt.Sprintf("Invalid log format: %s", settings.Format))
-		}
-		logger := &logrus.Logger{
-			Formatter: formatter,
-			Level:     level,
-			Out:       settings.Output,
-			Hooks:     make(logrus.LevelHooks),
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	currentSettings = settings
+	level, err := logrus.ParseLevel(settings.Level)
+	if err != nil {
+		panic(fmt.Sprintf("Could not parse log level configuration: %s", err))
+	}
+	formatter, ok := supportedFormatters[settings.Format]
+	if !ok {
+		panic(fmt.Sprintf("Invalid log format: %s", settings.Format))
+	}
+
+	output, ok := supportedOutputs[settings.Output]
+	if !ok {
+		panic(fmt.Sprintf("Invalid output: %s", settings.Output))
+	}
+
+	entry := ctx.Value(logKey{})
+	if entry == nil {
+		entry = defaultEntry
+	} else {
+		defaultEntry.Logger = &logrus.Logger{
+			Out:          output,
+			Hooks:        defaultEntry.Logger.Hooks,
+			Formatter:    formatter,
+			ReportCaller: defaultEntry.Logger.ReportCaller,
+			Level:        level,
+			ExitFunc:     defaultEntry.Logger.ExitFunc,
 		}
-		hook := filename.NewHook()
-		hook.Field = FieldComponentName
-		logger.AddHook(hook)
-		defaultEntry = logrus.NewEntry(logger)
-		defaultEntry = defaultEntry.WithField(FieldCorrelationID, "-")
-	})
-	return ContextWithLogger(ctx, defaultEntry)
+	}
+	e := entry.(*logrus.Entry)
+	e.Logger = &logrus.Logger{
```

Comment:
I remember that this caused a race condition in the past, because you need the same mutex from the entry's logger, but it's private

Reply:
not changing the logger causes a race condition currently

Reply:
seems to work without copying the logger after I changed one of the tests that failed with race, so we can keep the same logger for all entries